Feature/content link client routing

nicholasio opened the pull request:
**What**:

Introduces a processor to replace content links with the Link component.

**Why**:

It's needed to allow frontity to handle internal links automatically (without causing a new page load)

**How**:

A processor was created to replace all anchor tags that have an href with our Link component.

**Tasks**:

<!-- Have you done all of these things?  -->

<!-- To check an item, place an "x" in the box like so: "- [x] Unit tests" -->

<!-- Move any unrelated task to the Unrelated tasks section below. -->

- [x] Code
- [x] TSDocs
- [x] TypeScript
- [x] Unit tests
- [x] Open an issue for this feature in the docs repo: https://github.com/frontity/docs/issues/248
- [x] Update community discussions
- [x] Add a changeset (with link to its [Feature Discussion](https://community.frontity.org/c/33) if it exists)

<!-- Changesets are necessary if your changes should release any packages.
Run `npx changeset` to create a changeset.
More info at https://docs.frontity.org/contributing/code-contribution-guide#what-is-a-changeset -->

**Unrelated Tasks**

<!-- ignore-task-list-start -->
- [ ] TypeScript tests
- [ ] End to end tests
- [ ] Update starter themes
- [ ] Update other packages
<!-- ignore-task-list-end -->

**Additional Comments**

<!-- Feel free to add any additional comments. -->

nicholasio committed:
First pass at implementing links processor

nicholasio committed:
Merge remote-tracking branch 'upstream/dev' into dev

nicholasio committed:
Merge branch 'dev' into feature/content-link-client-routing

nicholasio committed:
Initial version of link processor

nicholasio committed:
Leave anchor tags without real links alone.

nicholasio committed:
wip

nicholasio committed:
Merge remote-tracking branch 'upstream/dev' into dev

nicholasio committed:
Merge remote-tracking branch 'upstream/dev' into dev

nicholasio committed:
Merge branch 'dev' into feature/content-link-client-routing

nicholasio committed:
Updating Link component to work with the new state.url setting

nicholasio committed:
Removing WpSource from packages

nicholasio committed:
replace the wp url for both http and https links

nicholasio committed:
Allowing electron brownser as it's useful for running cypress inside WSL 2

nicholasio committed:
Add changeset

nicholasio committed:
Add FD link to changeset

nicholasio committed:
Adding tests for link processor

nicholasio commented on e2e/e2e.js:
```diff
@@ -42,7 +42,9 @@ let isDockerRunning = false;
 validateArgs(target, { possibleValues: ["es5", "module", "both"] });
-validateArgs(browser, { possibleValues: ["firefox", "chrome", "edge"] });
+validateArgs(browser, {
+  possibleValues: ["firefox", "chrome", "edge", "electron"],
```
I'm not sure if there was a reason to leave `electron` out but I find it useful for being able to easily run tests in WSL 2 without having to install chrome. BTW the e2e stuff worked fine in WSL 2 with Docker for Desktop (Windows).

luisherranz replied:
No reason. We were simply not using it anymore but no problem in adding it back 🙂

luisherranz commented on .changeset/itchy-moles-enjoy.md:
```diff
@@ -0,0 +1,8 @@
+"@frontity/html2react": minor
+---
+
+Introduces a `replaceWPUrls` prop to the `Link` component and `link` processor that replaces all anchor tags with the `Link` component to make internal links work.
```
Could you please change the name from `replaceWPUrls` to `replaceSourceUrls`?

nicholasio replied:
Done.

luisherranz commented on packages/html2react/processors/link.tsx:
```diff
@@ -0,0 +1,35 @@
+  name: "link",
+  test: ({ node }) => node.component === "a",
+  processor: ({ node }) => {
+    if (node.props?.href && !node.props?.href?.startsWith("#")) {
```
Interesting that you are leaving out the hash links. @orballo could you please take a look at that logic once you make some progress with the [hash link feature](https://community.frontity.org/t/add-support-for-hash-links-in-frontity-tiny-router/3189)?

@nicholasio by the way, you can move that conditional to the `test` function.

nicholasio replied:
I'll move the conditional to the test function.

Regarding the hash links - not sure of the benefits of handling it through router/link component but I can certainly include hash links.

luisherranz replied:
I'm fine merging this as it is, I just wanted to draw the attention of orballo to this processor to make sure it works fine with his implementation 🙂

luisherranz commented on packages/html2react/src/index.ts:
```diff
@@ -1,12 +1,13 @@
   libraries: {
     html2react: {
-      processors: [script],
+      processors: [script, link],
```
I guess we have to make a decision here: should the `link` processor be added by default or should be added by the theme creator like the `image` and `iframe` processors?

My opinion now is that we should leave it out for the moment, as the `<Link>` component itself is optional in the themes. But we should improve the way people share processors. The conversation is happening here: https://community.frontity.org/t/how-to-share-html2react-processors/3069. It would be awesome if you could give us your feedback @nicholasio 🙂.

The best idea we had so far is to make them similar to the way `fills` work now. It is a bit verbose, but it is consistent and very flexible.

`script` is not optional because if not the `<script>` tags of the `post.content` don't work anymore.

nicholasio replied:
I was going to say that it makes sense but if the script is not optional it means that all themes are already shipping with frontity/components package, so I don't see a reason why not to make the link processor be enabled by default.

nicholasio replied:
Actually, I see your point, you're concerned more about the way processors are shared and not themes needing to declare `frontity/components` as a dependency. I'll make it optional and include in the default themes.

nicholasio replied:
Done.

luisherranz replied:
Thanks @nicholasio :)

luisherranz commented on packages/html2react/processors/__tests__/link.test.tsx:
```diff
@@ -0,0 +1,35 @@
+    const { container } = render(
+      <Html2React
+        html="<a href='http://my-link.com/post-name'>My Link</a><a href='http://my-link.com/post-name'>My Link 2</a>"
+        {...store}
```
@DAreRodz maybe we can simplify the processor tests now that you have added a `processors` prop to `Html2React`.

What do you think? Would something like this work?

```js
<Html2React
  html="<a href='http://my...ame'>My Link 2</a>"
  processors={[link]}
/>
```

DAreRodz replied:
Sure, that should work fine. When the `processors` prop is present, the content of `libraries.html2react.processors` is just ignored.

nicholasio replied:
FYI: TS complains about types not matching if we just pass processors since the other props aren't optional.

DAreRodz replied:
Oh, I see. That happens because the _disconnected_ `Html2React` is being imported

```ts
import { Html2React } from "../../src/libraries/component";
```
instead of the _connected_ version, which is the default export:

```ts
import Html2React from "../../src/libraries/component";
```

That one should not require the Frontity store props to be passed. :slightly_smiling_face:

nicholasio replied:
That worked, updated 👍

luisherranz commented on packages/components/link/utils.ts:
```diff
@@ -111,3 +111,17 @@ export const onHover = (el: HTMLAnchorElement, cb: () => void) => {
+ *
+ * @returns The URL without the WP URL.
+ */
+export const removeWPUrl = (link: string, wpURL: string) => {
```
Maybe it would be great to add some unit tests for this util? It would be easier to fix edge cases and it would make sure it's not messed up in future refactorings. What do you think @nicholasio?

nicholasio replied:
Yep, will add tests for this.

luisherranz replied:
Great. Thanks :)

nicholasio replied:
Done.

luisherranz reviewed (changes requested):
Hey, awesome work @nicholasio. Many thanks for this PR 🙂

The only thing I would change is `replaceWPUrls` to `replaceSourceUrls` because, although now only `wp-source` exists, theoretically the link can be used by any source, even sources other WordPress. Also, it relates more to the `state.source.url` setting, which is what uses underneath to figure out what to replace.

nicholasio committed:
CR Changes: Addting tests and renaming prop.

nicholasio committed:
Moving if check to the test function

nicholasio committed:
Making link processor optional and adding it to the default themes.

nicholasio committed:
Revert frontity.settings.ts

nicholasio committed:
Updating processors tests

luisherranz committed:
Refactor removeWPUrl and add tests for folders

luisherranz reviewed (approved):
Looks perfect to me! Thanks for the PR and thanks for refactoring also the other processor tests!

I have added tests for the case when `state.source.url` contains a folder and I have refactored the `removeWPUrl` function to make it work. Could you please review [the new implementation](https://github.com/frontity/frontity/pull/520/commits/8e9eac4ab65dd64d4e78975883d09390f4032d53#diff-a2a24789b184eb4ccec4bf21e40d3a76006682d94c7f95d56e76c8a3bd7445a3R124-R128) and let me know if that is how you think it should work? Once you give me your approval I will merge the PR 🙂

luisherranz commented:
BTW, it seems like an E2E test started failing but it's not related: https://github.com/frontity/frontity/issues/619

nicholasio commented:
> Looks perfect to me! Thanks for the PR and thanks for refactoring also the other processor tests!
> 
> I have added tests for the case when `state.source.url` contains a folder and I have refactored the `removeWPUrl` function to make it work. Could you please review [the new implementation](https://github.com/frontity/frontity/pull/520/commits/8e9eac4ab65dd64d4e78975883d09390f4032d53#diff-a2a24789b184eb4ccec4bf21e40d3a76006682d94c7f95d56e76c8a3bd7445a3R124-R128) and let me know if that is how you think it should work? Once you give me your approval I will merge the PR 🙂

Looks good to me! 👍

luisherranz committed:
Merge branch 'dev' into feature/content-link-client-routing

luisherranz commented:
Awesome. Merged now. Thanks again @nicholasio! 🎉
